staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go

```diff
@@ -31,171 +31,278 @@ import (
+func SetNestedField(obj map[string]interface{}, value interface{}, fields ...string) bool {
+	assertValueType(value)
 	m := obj
-	if len(fields) > 1 {
-		for _, field := range fields[0 : len(fields)-1] {
-			if _, ok := m[field].(map[string]interface{}); !ok {
-				m[field] = make(map[string]interface{})
+	for _, field := range fields[:len(fields)-1] {
+		if val, ok := m[field]; ok {
+			if valMap, ok := val.(map[string]interface{}); ok {
+				m = valMap
+			} else {
+				return false
 			}
-			m = m[field].(map[string]interface{})
+		} else {
+			newVal := make(map[string]interface{})
+			m[field] = newVal
+			m = newVal
 		}
 	}
 	m[fields[len(fields)-1]] = value
+	return true
 }
 
-func setNestedSlice(obj map[string]interface{}, value []string, fields ...string) {
+// SetNestedStringSlice sets the string slice value of a nested field.
+// Returns false if value cannot be set because one of the nesting levels is not a map[string]interface{}.
+func SetNestedStringSlice(obj map[string]interface{}, value []string, fields ...string) bool {
 	m := make([]interface{}, 0, len(value))
 	for _, v := range value {
 		m = append(m, v)
 	}
-	setNestedField(obj, m, fields...)
+	return SetNestedField(obj, m, fields...)
+}
+
+// SetNestedSlice sets the slice value of a nested field.
+// Returns false if value cannot be set because one of the nesting levels is not a map[string]interface{}.
+func SetNestedSlice(obj map[string]interface{}, value []interface{}, fields ...string) bool {
+	m := make([]interface{}, len(value))
+	copy(m, value)
+	return SetNestedField(obj, m, fields...)
+}
+
+// SetNestedStringMap sets the map[string]string value of a nested field.
+// Returns false if value cannot be set because one of the nesting levels is not a map[string]interface{}.
+func SetNestedStringMap(obj map[string]interface{}, value map[string]string, fields ...string) bool {
+	m := make(map[string]interface{}, len(value))
+	for k, v := range value {
+		m[k] = v
+	}
+	return SetNestedField(obj, m, fields...)
 }
 
-func setNestedMap(obj map[string]interface{}, value map[string]string, fields ...string) {
+// SetNestedMap sets the map[string]interface{} value of a nested field.
+// Returns false if value cannot be set because one of the nesting levels is not a map[string]interface{}.
+func SetNestedMap(obj map[string]interface{}, value map[string]interface{}, fields ...string) bool {
 	m := make(map[string]interface{}, len(value))
 	for k, v := range value {
 		m[k] = v
 	}
-	setNestedField(obj, m, fields...)
+	return SetNestedField(obj, m, fields...)
+}
+
+// RemoveNestedField removes the nested field from the obj.
+func RemoveNestedField(obj map[string]interface{}, fields ...string) {
+	m := obj
+	for _, field := range fields[:len(fields)-1] {
+		if x, ok := m[field].(map[string]interface{}); ok {
+			m = x
+		} else {
+			return
+		}
+	}
+	delete(m, fields[len(fields)-1])
+}
+
+func getNestedString(obj map[string]interface{}, fields ...string) string {
```

Comment:
nit: nestedStringOrDefault with `default string` arg?

Reply:
`getNestedString()` is used in lots of places and default is an empty string everywhere. That is why I kept it as it is. Also I didn't want to touch the name because it would've increased the PR size even further. I can send a follow-up PR with rename if needed.

Reply:
Makes sense.